README.md
SIMcheck
========

SIMcheck is a package of ImageJ tools for assessing the quality and
reliability of Structured Illumination Microscopy (SIM) data.

* More information can be found on the
[Micron Oxford Website](http://www.micron.ox.ac.uk/software/SIMCheck.php)
* Further help is available
[here](http://www.micron.ox.ac.uk/microngroup/software/SIMcheck.html)

Copyright Graeme Ball and Lothar Schermelleh, Micron Oxford, Department of
Biochemistry, University of Oxford. License GPL unless stated otherwise in
a given file (in particular, SIMcheck uses modified versions of ImageJ's
Slicer plugin and Paul Baggethun's Radial Profile Plot plugin).


Installation
============

ImageJ updater
--------------

The simplest way to install SIMcheck and keep it up to date on your system is
using the ImageJ updater.  SIMcheck is available via an ImageJ update site so
you only need to activate it from the list of sites.  See
["How to follow an update site"](http://fiji.sc/How_to_follow_a_3rd_party_update_site)
for more details

Manual install
--------------

If for some reason the ImageJ updater is not available, the jar file
for the latest release can still be [downloaded
manually](http://downloads.micron.ox.ac.uk/fiji_update/SIMcheck/plugins/)
from the update site.  If so, the file extension needs to be manually
adjusted adn dependencies will need to be installed somehow.

Building from source
--------------------

The project uses the maven build and dependency management tool, so to
build it, run the following command (.jar file appears in `target/`):

    mvn package

In addition, SIMcheck is also released via Sonatype to ease its use by
other projects:

    <dependency>
      <groupId>uk.ac.ox.micron</groupId>
      <artifactId>SIMcheck</artifactId>
      <version></version>
    </dependency>


Features
========

0: SIMcheck main dialog
-----------------------

- dialog to choose and set up all checks with standard parameters
- cropping utility for raw and reconstructed data
- results: images will appear and key statistics will be logged
- help button: link to instructions, help, documentation

1: Pre-processing, Raw Data Checks
----------------------------------

    Check            |        Statistic(s)                 |      Comments
-------------------- | ----------------------------------- | ------------------
 Intensity Profiles  | bleaching, flicker, angle intensity |
 Motion / Illum Var  | None: angle differences colored     |
 Fourier Projections | None: check pattern / spots OK      |
 Modulation Contrast | feature MCNR acceptable?            | Wiener estimate

2: Post-reconstruction Checks
-----------------------------

    Check            |        Statistic(s)                 |      Comments
-------------------- | ----------------------------------- | ------------------
 Intensity Histogram | +ve/-ve ratio acceptable?           | top/bottom 0.01%
 SA Mismatch         | stdDev of miniumum vs. mean         | shows OTF mismatch
 Fourier Plots       | None: symmetry+profile OK?          |
 Mod Contrast Map    | None: inspect MCNR of features      | green=saturated

3: Calibration Checks
---------------------

    Check            |        Statistic(s)                 |      Comments
-------------------- | ----------------------------------- | ------------------
 Illum. Phase Steps  | phase step & range stable?          | +k0, linespacing
 Pattern Focus       | None: check for "zipper" pattern    |

4: Utilities
------------

- Format Converter for SIM data
- Raw SIM Data to Pseudo-Widefield conversion
- Threshold and 16-bit conversion (i.e. "discard negatives")
- Stack FFT, performs 2D FFT on each slice


PROJECT STRUCTURE
=================

- README.md - This file
- LICENSE.txt - Project's license
- CHANGES.txt  - History of versions and changes
- pom.xml - maven Project Object Model describing dependencies, build etc.
- src/main/java/ - Application sources (in SIMcheck/ package)
- src/main/resources/ - Application resources (IJ menu config, html help text)
- src/test/java/ - Test sources
- src/test/resources/ - Test resources
- target/ - output `SIMcheck_.jar` file
- target/classes/ - build output .class files
- target/test-classes/ - classes produced by tests


Style Notes (1.0)
===========

* no run-time dependencies other than ImageJ1
* simple, modular structure - each check is a standalone plugin
* plugin exec methods take input images and return ResultSet
  (no GUI calls within exec when easily avoidable)
* ImageJ1-like preference for pre- java 5 features and reliance on float
  primitive type for most calculations


TODO
====

* 1.2: post-release updates, bugfixes & refactoring

    - MCN: report per. angle MCNR & warn about saturated pixels
    - Rec Fourier & Stack FFT utility: tidy options, use RadioButtons
    - si2wf: add option to select 1 angle
    - raw FPJ: add target overlay? (hide by default)
    - progress bar for FPJ plugin (& others?)
    - MCN, show saturated pixels in raw data? (add to LUT?)
    - Rec Fourier: improve 3D FFT dependency (maven dep: IJ2? JTransforms?)
    - more crop utility tests, move to separate utility plugin
    - turn FTR profile into multi-color and/or plot
    - improve "Fourier Transform Phases" info / log output
    - angle labels etc. should be overlaid, not drawn

* 1.3: PSF and OTF symmetry, extent, shape & order separation

* 1.4: additional numerical stats, including resolution estimate

    - stat for motion & illumination variation
    - SIR Fourier: resolution estimate? stat for artifacts?
    - multi-frame: test / document stats for current time-point only?

* 1.5: ImageJ2 reimplementation / headless running

    - get rid of IJ.run calls & intermediate results from window list
    - documentation: macro examples / document batch running

* 1.6: check name updates & test suite

    - Rec Fourier: rename ortho to axial?
    - better name for spherical aberration mismatch?
    - expand sample / test data set
    - optional tests using failsafe plugin? (check data & pass/fail)

* 2.0: integrated swing GUI control


SIM Reconstruction Problems & Remedies
--------------------------------------
(Kai Wicker)

- **Camera background**: subtraction

- **optical distortions & uneven gain**: flat-field correction

- **illumination intensity fluctuations**: normalisation (after background subtraction)

- **sample drift between images**: drift correction by means of cross-correlation

- **unknown grating period**: find grating vector using component cross-correlations

- **fluctuations in grating phase**: optimization of mixing matrix

- **fluctuation of order strengths between images**

- **sample drift between focal slices**: drift correction through cross-correlation

- **unknown zero grating phase**: global phase via cross-correlation

- **drift between rotational orientations**: drift correction through cross-correlation

- **unknown order strengths**: compare different separated components


CHANGES.txt
Version 0.8
-----------
* significant speed-up of mod contrast calc
* fixed several problems with reconstructed Fourier plots
* improved Fourier plots (window function, calibration)
* fixed mod contrast map display scaling
* ELYRA data conversion from main dialog now works
* better log info - tidied output, preparation for GUI

Version 0.9
-----------
* SI "Z variation" plugin now normalized to average values
  in the image, making it more robust
* Modulation contrast calculation improved / fixed,
** so it is valid at different noise levels 
** uses illumination pattern orders 1 and 2 instead of 1
** a noticeable effect is that typical mod contrast values 
   are halved
* intensity histogram and mismatch plugins produce stacks
* reconstructed Fourier check produces a radial profile plot
* phase calibration plugin improved,
** detects peaks from pattern order 1 and plots them
** estimates line spacing and k angle from the peaks
** produces result table for phase step analysis
* new N-SIM format converter

Version 0.9.5
-------------
* new SIMcheck Manual (using the Sphinx documentation tool)
* fixes to Phase Step calibration check (Ian Dobbie)
* switched to Apache Maven for build & dependency management
* reconstructed Fourier Plots streamlined and improved
* many of the checks have been renamed
* fixed bug in Reconstructed Intensity Histogram statistic
* code tidied and more tests added, several minor bugs fixed
* Modulation Contrast Map shows saturation in raw data (green)
* new Illumination Pattern Focus calibration check
* new Threshold and 16-bit Conversion utility
* raw data Fourier plugin now displays montage of projections
* fixes to ELYRA and N-SIM data format conversion
* implemented image-cropping utility in main SIMcheck dialog
* added option to specify camera bit depth (to identify saturation)
* Look-Up Tables displayed as overlay on results when appropriate
* the log output / report generation has been tidied up

Version 0.9.6
-------------
* new Result Table summarising key numerical stats and interpretation
* RMSE stat for MIV check implementation finished
* CIP changed to report overall intensity fluctuation statistic

Version 0.9.7
-------------
fixed many minor bugs, the most important of which are:
* preserve calibration info in all checks
* several bugs in SI to pseudo-widefield utility
* incorrect FTR radial profile scale
* show saturation in MCM if *any* raw data angle saturated (not all as before)
other updates and notable new features:
* added check for ZMV value
* reformatted and tidied log output, plot axes and titles
* added Results summary table with stat value checks
* SI to pseudo-widefield: preserve data type, scale XY x2 to match super-res
* added flicker correction routine to IPF cabliration check and PWF utility
* improved intensity profile stats, inlcluding total variation and flicker
* improved stack feature intensity estimates (now using Otsu method)
* raw fourier projection now projects all angles for 1 summary image
* increased default FFT window function size
* raw FPJ and reconstructed SAM checks default off in main dialog
* change SA mismatch calibration check back to reconstructed

Version 0.9.8
-------------
* extensive SIMcheck Manual updates, with figures for visual guide
* feature: Stack FFT 2D gamma scaling display option (reviewer request)
* add '_' to jar finalName via pom so plugin appears in IJ Plugins
* bugfix: si2wf accidental rescale
* Cal_PatternFocus change default angle def method to line

Version 0.9.9
-------------
* bugfix: Util_RescaleTo16bit incorrect minimum
* add new FFT2D display option (log-scaled, float) and refactor
* update all javadocs

Version 0.9.10
--------------
* IPS, set plot width to ensure it fits stack of plots
* report all stats to 3 sig fig, previously 2
* pattern focus, add CCW stripe rotation checkbox
* change default pattern focus angle method back to IJ degrees
* CIP bugfix: was overwriting intens array for each C (pass ref to Plot)
* 16-bit util: fix for missing subtraction of specified min

Version 0.9.11
--------------
* make reconstructed FFT gamma 0.2 display default

Version 1.0.0
-------------
* change raw MCN stat check criteria from 3,6 to 4,8
* python script for dumping binary LUT file as comma-separated text
* rescale to 16-bit in Stack FFT utility to match reconstructed FFT
* bugfix: I1l.getStatsForChannel and affected display settings
* raw FPJ: now uses 32-bit log and central mask, display min-max
* raw CIP check criteria change from 50,70 to 60,40
* CIP: improve bleach estimation, add normalised plot
* updates to allow macro recording

Version 1.1
-----------
* SI to widefield utility: add option to skip XY rescale 2x
* implement 3D FFT using Parallel FFTJ for Rec Fourier check
* "FTR" radial profile plot: fixed range 20-40 for log-scaled PS

Version 1.2
-----------
* Fix URL to online help

Version 1.3
-----------
* New "Modulation Contrast Filter" utility.

Version 1.4
-----------


src/main/java/SIMcheck/ResultSet.java
/* Copyright (c) 2015, Graeme Ball.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/ .
 */

package SIMcheck;

import java.util.*;

import ij.IJ;
import ij.ImageJ;
import ij.ImagePlus;
import ij.measure.ResultsTable;

/** ImageJ1 plugin result container using HashMaps: items must have
 * unique names within result types.
 * @author Graeme Ball <graemeball@gmail.com>
 */
public class ResultSet {

    // for automatic formatting of result log / output
    private static final int TEXTWIDTH = 52;
    static final int STAT_SIG_FIGS = 3;
    private static final int CHECK_MAX_CHARS = TEXTWIDTH * 100;

    /** Interpretation of a statistic: is it OK? yes, no, maybe. */
    public enum StatOK {

        YES("Yes"),
        NO("No"),
        MAYBE("?"),
        NA("N/A"); // stat has no interpretation -- for info only

        private String desc;
        StatOK(String desc) {
            this.desc = desc;
        }

        public String str() {
            return desc;
        }
    }

    /** Store a numerical statistic and its interpretation. */
    private class Stat {
        Double value;
        StatOK statOK;
        Stat(Double value, StatOK statOK) {
            this.value = value;
            this.statOK = statOK;
        }
    }

    String resultSetName = "";
    String resultSetTLA = "";
    private LinkedHashMap<String, ImagePlus> imps =
            new LinkedHashMap<String, ImagePlus>();
    private LinkedHashMap<String, Stat> stats =
            new LinkedHashMap<String, Stat>();
    private LinkedHashMap<String, String> infos =
            new  LinkedHashMap<String, String>();

    ResultSet(String name, String tla) {
        resultSetName = name;
        resultSetTLA = tla;  // TODO: store ref to ePlugin class instead
    }

    /** Add ImagePlus result & description: title+description MUST be unique. */
    public void addImp(String description, ImagePlus imp) {
        // key composed of image title + description -- separated again later
        description = imp.getTitle() + ":\n" + description;
        if (imps.containsKey(description)) {
            throw new IllegalArgumentException(description + " already exists");
        }
        imps.put(description, imp);
    }

    /** Return Imp number nImp (insert order, 0+) in imps hashmap, else null */
    public ImagePlus getImp(int nImp) {
        Iterator<ImagePlus> it = imps.values().iterator();
        int n = 0;
        while (it.hasNext()) {
            if (n == nImp) {
                return it.next();
            }
        }
        return null;  // did not reach desired nImp
    }

    /**
     * Add a named Double statistic result: statName MUST be unique.
     * NB: it is the plugin's responsibility to decide whether stat is OK.
     */
    public void addStat(String statName, Double value, StatOK statOK) {
        if (stats.containsKey(statName)) {
            throw new IllegalArgumentException(statName + " already exists");
        }
        stats.put(statName, new Stat(value, statOK));
    }

    /** Add an information string with title: title MUST be unique. */
    public void addInfo(String title, String info) {
        if (infos.containsKey(title)) {
            throw new IllegalArgumentException(title + " already exists");
        }
        infos.put(title, info);
    }

    /** Report all results to ImageJ log window. */
    public void report() {
        StringBuilder sb = new StringBuilder(CHECK_MAX_CHARS);
        sb.append("\n");
        sb.append(J.nChars((int)(TEXTWIDTH * 1.0), "-") + "\n");
        sb.append(titleString(resultSetName, " ") + "\n");
        sb.append(J.nChars((int)(TEXTWIDTH * 1.0), "-") + "\n");
        for (Map.Entry<String, ImagePlus> entry : imps.entrySet()) {
            String description = entry.getKey();
            ImagePlus imp = (ImagePlus)entry.getValue();
            sb.append("Displaying " + imp.getTitle() + ":\n");
            int nTitleChars = imp.getTitle().length() + 2;
            description = description.substring(
                    nTitleChars, description.length());
            sb.append(autoFormat(description, (int)(TEXTWIDTH * 1.15), 0));
            sb.append("\n");
            imp.show();
        }
        if (!stats.isEmpty()){
            sb.append("\nStatistics:\n");
        }
        // loop over stats twice: log checked (non-NA) stats first, rest after
        boolean hasCheckedStats = false;
        for (Map.Entry<String, Stat> entry : stats.entrySet()) {
            String statName = entry.getKey();
            if (stats.get(statName).statOK != StatOK.NA) {
                sb.append(statName + " = " + J.d2s(entry.getValue().value));
                sb.append("\n");
                hasCheckedStats = true;
            }
        }
        if (hasCheckedStats) {
            sb.append("--\n");  // print a separator if something to separate
        }
        for (Map.Entry<String, Stat> entry : stats.entrySet()) {
            String statName = entry.getKey();
            if (stats.get(statName).statOK == StatOK.NA) {
                sb.append(statName + " = " + J.d2s(entry.getValue().value));
                sb.append("\n");
            }
        }
        for (Map.Entry<String, String> entry : infos.entrySet()) {
            String infoTitle = entry.getKey();
            String info = entry.getValue();
            if (info.length() > 20) {
                sb.append("\n");  // blank line between infos unless very short
            }
            String sep = ": ";
            if (infoTitle.charAt(infoTitle.length() - 1) == '!') {
                sep = " ";  // don't add a colon if title ends on exclam mark!
            }
            sb.append(infoTitle + sep + autoFormat(info, (int)(TEXTWIDTH * 1.15),
                    infoTitle.length() + 2));
            sb.append("\n");
        }
        IJ.log(stripMultipleBlankLines(sb.toString()));
    }

    /**
     * Display summary Result table of numerical stats and interpretation
     * for all results in a list of ResultSets.
     */
    public static void summary(List<ResultSet> resultSets, String title) {
        ResultsTable rt = new ResultsTable();
        for (ResultSet rs : resultSets) {
            for (String statName : rs.stats.keySet()) {
                // report all stats except where interpretation N/A
                if (rs.stats.get(statName).statOK != StatOK.NA) {
                    rt.incrementCounter();
                    rt.addValue("Check", rs.resultSetTLA);
                    rt.addValue("Statistic", statName);
                    rt.addValue("Value", J.d2s(rs.stats.get(statName).value));
                    rt.addValue("Pass", rs.stats.get(statName).statOK.str());
                }
            }
        }
        rt.show(title);
    }

    /**
     * Automatically add line-breaks, returning text String of fixed width.
     * titleLen shortens the first line, adjusting for length of item title.
     * '  -' starts a new indented list item
     * lists *must* be terminated with '  -- '
     */
    private static String autoFormat(String text, int width, int titleLen) {
        StringBuilder sb = new StringBuilder((int)(text.length() * 1.1));
        int thisSpace = 0;
        int lastSpace = 0;
        int lineStart = 0;
        int maxIter = 100;  // prevent infinite while; implies < maxIter words
        int iter = 0;
        boolean firstLine = true;
        boolean insideList = false;
        boolean hasItemList = false;
        while (thisSpace != -1 && iter < maxIter) {
            lastSpace = thisSpace;
            thisSpace = text.indexOf(" ", lastSpace + 1);
            if (thisSpace == -1) {
                sb.append(text.substring(lineStart, text.length()));
            } else if (thisSpace + 4 < text.length() - 1 &&
                    text.substring(thisSpace, thisSpace + 4).equals("  - ")) {
                insideList = true;
                // handle indentation of list items starting '  -'
                sb.append(text.substring(lineStart, thisSpace) + "\n");
                sb.append("- ");
                lineStart = thisSpace + 4;
                thisSpace += 4;
            } else if (thisSpace + 4 < text.length() &&
                    text.substring(thisSpace, thisSpace + 4).equals("  --")) {
                // end multi-line lists upon ' --'
                hasItemList = true;
                insideList = false;
            } else {
                int adjustedLineStart = lineStart;
                if (firstLine) {
                    // first line, with title, would be too long w/o adjustment
                    adjustedLineStart -= titleLen;
                }
                if (thisSpace - adjustedLineStart > width) {
                    // backtrack to lastSpace
                    sb.append(text.substring(lineStart, lastSpace) + "\n");
                    lineStart = lastSpace + 1;
                    if (insideList) {
                        sb.append("   ");  // extra indent for list items
                    }
                    if (firstLine) {
                        firstLine = false;
                    }
                }
            }
            iter++;
        }
        if (insideList == true) {
            throw new IllegalArgumentException(
                    "malformed text for autoformatting:\n" +
                    " item lists should be terminated with '  -- '");
        }
        if (hasItemList) {
            // trim off "  --" before returning
            return sb.toString().substring(0, sb.length() - 4);
        } else {
            return sb.toString();
        }
    }

    /** Return an Object[] representation of the results. */
    public Object[] objects() {
        List<Object> objList = new ArrayList<Object>();
        objList.add(resultSetName);
        objList.addAll(Arrays.asList(imps.keySet().toArray()));
        objList.addAll(Arrays.asList(imps.values().toArray()));
        objList.addAll(Arrays.asList(stats.keySet().toArray()));
        objList.addAll(Arrays.asList(stats.values().toArray()));
        objList.addAll(Arrays.asList(infos.keySet().toArray()));
        objList.addAll(Arrays.asList(infos.values().toArray()));
        Object[] objArray = objList.toArray();
        return objArray;
    }

    /** Return a String containing 'title' centered in a line of charX. */
    static String titleString(String title, String charX) {
        // trial & error formatting tweaks due to non-fixed-width font :-(
        double textLen = title.length() * 0.83;
        if (!title.equals("")) {
            title = " " + title + " ";
        }
        int nPadChars = (int)(0.5 * (TEXTWIDTH - textLen));
        // char width corrections for IJ's non-fixed-width font
        if (charX.equals("=")) {
            nPadChars = (int)(0.72 * nPadChars);  // correct for wider '='
            if (title.equals("")) {
                nPadChars++;  // minor tweak
            }
        } else if (charX.equals(" ")) {
            nPadChars = (int)(1.9 * nPadChars);  // correct for narrow ' '
        } else if (charX.equals("-")) {
            nPadChars = (int)(0.9 * nPadChars);  // correct for '-' width
        }
        return J.nChars(nPadChars, charX) + title + J.nChars(nPadChars, charX);
    }

    /** Replace occurrences of multiple blank lines with 1 blank line. */
    private static String stripMultipleBlankLines(String s) {
        return s.replaceAll("\n{3,}", "\n\n");
    }

    /** test method */
    public static void main(String[] args) {

        ResultSet results = new ResultSet("ResultSet Test", "TST");
        new ImageJ();
        ImagePlus blimp = TestData.lawn;
        results.addImp("a bead lawn", blimp);
        results.addStat("stat1, imWidth",
                (double)blimp.getWidth(), StatOK.YES);
        results.addStat("stat2, imHeight",
                (double)blimp.getHeight(), StatOK.YES);
        results.addStat("stat3, imBytesPerPix",
                (double)blimp.getBytesPerPixel(), StatOK.YES);
        results.addInfo("about", "this is raw SIM data for a bead lawn");
        results.addInfo("info list", "list items auto-formatted:"
                + "  - item 1  - item 2  - item 3"
                + "  - item 4: long item exceeding TEXTWIDTH that tests"
                + " multi-line wrap indentation"
                + "  -- ");  // '  -- ' teminated

        IJ.log("report()  - should show all images and log"
                + " all stats, info. Check stats appear in order.");
        results.report();

        IJ.log("\nchecking identity of " + results.objects().length
                + " (expected " + 11 + ") returned objects...");
        Object[] objs = results.objects();
        IJ.log("resultSetName, objs[0] instanceof String? "
                + new Boolean(objs[0] instanceof String).toString());
        IJ.log("imps key 1, objs[1] instanceof String? "
                + new Boolean(objs[1] instanceof String).toString());
        IJ.log("imps value 1, objs[2] instanceof ImagePlus? "
                + new Boolean(objs[2] instanceof ImagePlus).toString());
        IJ.log("stats key 1, objs[3] instanceof String? "
                + new Boolean(objs[3] instanceof String).toString());
        IJ.log("stats key 2, objs[4] instanceof String? "
                + new Boolean(objs[4] instanceof String).toString());
        IJ.log("stats key 3, objs[5] instanceof String? "
                + new Boolean(objs[5] instanceof String).toString());
        IJ.log("stats value 1, objs[6] instanceof Double? "
                + new Boolean(((Stat)objs[6]).value instanceof Double).toString());
        IJ.log("stats value 2, objs[7] instanceof Double? "
                + new Boolean(((Stat)objs[7]).value instanceof Double).toString());
        IJ.log("stats value 3, objs[8] instanceof Double? "
                + new Boolean(((Stat)objs[8]).value instanceof Double).toString());
        IJ.log("infos key 1, objs[9] instanceof String? "
                + new Boolean(objs[9] instanceof String).toString());
        IJ.log("infos value 1, objs[10] instanceof String? "
                + new Boolean(objs[10] instanceof String).toString());

        IJ.log("addStat() duplicate throws IllegalArgument exception?");
        try {
            results.addStat("stat1, imWidth", 7.0d, StatOK.YES);
            IJ.log("NO EXCEPTION");
        } catch (Exception e) {
            if (e instanceof IllegalArgumentException) {
                IJ.log("YES");
            } else {
                IJ.log("WRONG EXCEPTION");
            }
        }
    }
}
